tests/kafkatest/tests/core/fetch_from_follower_test.py

```diff
@@ -0,0 +1,145 @@
+        selector is set to the rack-aware implementation. One of the brokers has a different rack than the other two.
+        We then use a console consumer with the "client.rack" set to the same value as the differing broker. After
+        producing some records, we verify that the client has been informed of the preferred replica and that all the
+        records are properly consumed.
+        """
+
+        # Find the leader, configure consumer to be on a different rack
+        leader_node = self.kafka.leader(self.topic, 0)
+        leader_idx = self.kafka.idx(leader_node)
+        non_leader_idx = 2 if leader_idx != 2 else 1
+        non_leader_rack = "rack-b" if leader_idx != 2 else "rack-a"
+
+        self.logger.debug("Leader %d %s" % (leader_idx, leader_node))
+        self.logger.debug("Non-Leader %d %s" % (non_leader_idx, non_leader_rack))
+
+        self.producer = VerifiableProducer(self.test_context, self.num_producers, self.kafka, self.topic,
+                                           throughput=self.producer_throughput)
+        self.consumer = ConsoleConsumer(self.test_context, self.num_consumers, self.kafka, self.topic,
+                                        client_id="console-consumer", group_id="test-consumer-group-1",
+                                        consumer_timeout_ms=60000, message_validator=is_int,
+                                        consumer_properties={"client.rack": non_leader_rack, "metadata.max.age.ms": "3000"})
+
+        # Start up and let some data get produced
+        self.start_producer_and_consumer()
+        time.sleep(5)
+
+        consumer_node = self.consumer.nodes[0]
+        consumer_idx = self.consumer.idx(consumer_node)
+        jmx_attribute = "preferred-read-replica"
+
+        jmx_object_name = "kafka.consumer:type=consumer-fetch-manager-metrics,client-id=%s,topic=%s,partition=%d" % \
+                  ("console-consumer", self.topic, 0)
+        self.jmx_tool.jmx_object_names = [jmx_object_name]
+        self.jmx_tool.jmx_attributes = [jmx_attribute]
+        self.jmx_tool.start_jmx_tool(consumer_idx, consumer_node)
+
+        # Wait for at least one interval of "metadata.max.age.ms"
+        time.sleep(5)
+
+        # Read the JMX output
+        self.jmx_tool.read_jmx_output(consumer_idx, consumer_node)
+
+        all_captured_preferred_read_replicas = defaultdict(int)
+        self.logger.debug(self.jmx_tool.jmx_stats)
+
+        for ts, data in self.jmx_tool.jmx_stats[0].items():
+            for k, v in data.items():
+                self.logger.debug("JMX => %s %s" % (k, v))
+                if k.endswith(jmx_attribute):
+                    all_captured_preferred_read_replicas[int(v)] += 1
+
+        self.logger.debug("Saw the following preferred read replicas %s",
+                          dict(all_captured_preferred_read_replicas.items()))
+        for idx in (1, 2, 3):
```

Comment:
This loop is a little puzzling to me. Is it not the same as the following:

```python
assert all_captured_preferred_read_replicas[non_leader_idx] > 0, ...
```

Reply:
Heh, this is another hold over from the earlier version of this test -- i'll simplify